Immediately close with INVALID_TOKEN (#3107)
quicwg/base-drafts: draft-ietf-quic-transport.md

use, a server can cause the connection to be routed to a server instance with
more resources available for new connections.

If a server receives a client Initial that can be unprotected but contains an
invalid Retry token, it knows the client will not accept another Retry token.
The server can drop such a packet and allow the client to time out to detect

> handshake failure, but that is a significant latency penalty to the client.

Reviewer: Maybe instead:

> The server could discard such a packet, but this could result in the client continuing to send the invalid Initial packet until it is eventually forced to abandon the connection attempt.  A server that requires address validation SHOULD immediately close ({{imm...}}) the connection with an INVALID_TOKEN error to allow the client to detect the error sooner.  Note that a server has not established any state for the connection at this point and so does not enter the closing period.  Alternatively, a server MAY proceed with the connection without verifying the token, though the server MUST NOT consider the client address to be valid.

Reviewer 2: I definitely like this text better. That said, I'm not sure that we should recommend sending a connection close. Esp. when under DoS, we do not want servers doing more work than necessary.

Reviewer: What does Roberto say again?  I think that if the server is under duress, then dropping the packet is fine.  This is a recommendation that allows the server to help the client, that's all.  Do you think we need to acknowledge the possibility that the server is overloaded and add the third case?

Reviewer 2: The text seems to assume that an implementer will handle the third case, where I was leaning towards being explicit. Perhaps something like: "When under duress, the server could discard such a packet. Dropping such a packet could however result in ..." That makes it a bit more explicit.

Author: I reworked Martin's suggestions, though I preferred to keep the sentence I had to his first sentence.  

In regards to "When under duress, the server could discard such a packet." I don't think that's adding anything on top of being clear about the consequences of discarding it, so I'd rather not add it.

Reviewer: Discarding is mentioned up front, so we don't need to concern ourselves with repeating that.

use, a server can cause the connection to be routed to a server instance with
more resources available for new connections.

If a server receives a client Initial with an invalid Retry token,

> it knows the client will not accept another Retry token.  It can either

Reviewer 2: I don't understand this sentence. How would the server know this for sure?

Reviewer 3: That's a keen observation (I think).

There is a chance where a Retry token is considered invalid due to packet becoming corrupt while being transmitted. A server can only say for sure that "the client will not accept another Retry token" unless either of the following conditions are met:
* The token is accompanied by a checksum (or an integrity protection). The checksum is correct, but the contents of the Retry token is invalid.
* The server successfully unprotects the Initial.

Reviewer 4: The client could also connect to an unexpected endpoint or the token could have expired.

Reviewer 3: Let me retract my [comment above](https://github.com/quicwg/base-drafts/pull/3107/files#r341481158). We do not need to care about the packet becoming corrupt on the wire, because the prerequisite for responding with a CONNECTION_CLOSE is that the Initial that has been received can be unprotected.

To avoid confusion, maybe change "If a server receives a client Initial with an invalid Retry token" to "If a server receives a client Initial that can be unprotected but contains an invalid Retry token"?

> The client could also connect to an unexpected endpoint or the token could have expired.

I'd say that that's a failure in the server deployment. We do not discuss how that should be handled.

Reviewer 4: > I'd say that that's a failure in the server deployment. We do not discuss how that should be handled.

Since Retry is processed before handshake, why can't it just be the client sending to the wrong IP address?

Author: #3128 requires being able to distinguish between the two types of tokens.  I updated the text with Kazuho's suggestion about being able to unprotect the Initial.

It can either proceed with the handshake without verifying the token or

> immediately close ({{immediate-close}}) the connection with a connection

Reviewer 2: ```suggestion
immediately close ({{immediate-close}}) the connection with an INVALID_TOKEN
```

> error of INVALID_TOKEN to cause the handshake to fail quickly instead of

Reviewer 2: ```suggestion
error to cause the handshake to fail quickly instead of
```

> waiting for the client to timeout.

Reviewer 2: This suggests 2 options at the server, when there are 3: it can simply drop the packet and let the client time out. You may want to call out all three options. 

Suggestion: "If a server receives a client Initial that can be unprotected but contains an invalid Retry token, it knows the client will not accept another Retry token. The server can drop such a packet and allow the client to time out to detect loss of this connection, but that is a significant latency penalty to the client. If possible, the server SHOULD either immediately close the connection with an INVALID_TOKEN error or proceed with the handshake without verifying the token."

Reviewer 3: I agree that listing all the three options would be an improvement.

At the same time, I have a concern regarding the suggestion of doing either an immediate close or proceeding the handshake. Those two options both require the server to retain state that cannot be delayed by providing a Retry to token to the client.

I think it we might want to downgrade the SHOULD to MAY, or, suggest that a server can send a CONNECTION_CLOSE frame (with INVALID_TOKEN) without retaining state.

Reviewer 2: It should be possible to detect an invalid token without per-connection state, right? 

That said, I realized that I don't think we should recommend that a server that might be getting DoSsed SHOULD send out another packet in response. How about "A server that is able to maintain enough state MAY either ..."

Reviewer 3: By state I mean the closing and the draining state that gets built and retained once you invoke {immediate-close}. We should not recommend such behavior as it would create a DoS vector. But if it is not going to be a SHOULD but instead “can” or MAY, I think we can be precise and therefor the text we have now is fine.

Author: I added a sentence about not creating any state in this case.  PTAL.

use, a server can cause the connection to be routed to a server instance with
more resources available for new connections.

> If a server receives a client Initial with a unverifiable Retry token,

Reviewer 3: Maybe s/unverifiable/invalid/?

To me it seems that "unverifiable" might have the implication of the decryption process (or the integrity check) of the token failing. But a rejection of Retry token happens in other cases too, e.g., when the Retry token is outdated.

Author: Done.

it knows the client will not accept another Retry token.  It can either

> proceed with the handshake without verifying the token or immediately close

Reviewer: I don't think that we should allow the server to continue.  Why not make the INVALID_TOKEN mandatory?

Reviewer 3: If we are to make this a mandatory requirement, under the assumption that the servers would act like that, should we also suggest that a server should issue tokens in a way that a Retry token and a NEW_TOKEN token are distinguishable from each other?

Unless we recommend servers building a token in that way, I do not think making this a MUST has benefits.

Reviewer 5: If we assume that servers only require Retries when they’re under attack, they might not bother checking the validity of a token when they’re not under attack.

Reviewer: > A token SHOULD be constructed in a way that allows the server to distinguish it from tokens that are sent in Retry packets as they are carried in the same field.

-- [Section 8.1.2](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#validate-future)

We could strengthen that to MUST if you like.  It probably should be anyway.

Author: The server could decide it's not under heavy load and proceed with the handshake, even if the token doesn't constitute address validation.  ie: The token could expire on a shorter timeline than the session ticket/etc, so 0-RTT could succeed, but address validation would fail.

Reviewer: I'm persuaded by Marten's comment here (which I missed at first, I think because we commented at almost the same time).

Reviewer 6: I still think the requirement to make them distinguishable should be a MUST.  Otherwise, a change of token generation method could make an old NEW_TOKEN token appear to be an invalid Retry token.

({{immediate-close}}) the connection with an connection error of
INVALID_TOKEN to cause the handshake to fail quickly instead of waiting
for the client to timeout.  When a client receives a CONNECTION_CLOSE

> with an error of INVALID_TOKEN, it MAY create a new connection.

Reviewer: This isn't a very useful "MAY", because we couldn't stop a client from doing this.  We never say anything about making other connections anywhere else.

Author: SG, sentence removed.

use, a server can cause the connection to be routed to a server instance with
more resources available for new connections.

If a server receives a client Initial that can be unprotected but contains an

> invalid Retry token, it knows the client will not accept another Retry token.

Reviewer 5: This only applies if the token is a Retry token (and not a NEW_TOKEN token).
When the client uses a NEW_TOKEN token, it would exchange this token for the Retry token when it receives a Retry packet, wouldn't it?

Reviewer: Note that the text explicitly says "Retry token", so I don't see how this is a problem.

The server can discard such a packet and allow the client to time out to

> detect handshake failure, but that is a significant latency penalty to the

Reviewer: ```suggestion
detect handshake failure, but that could impose a significant latency penalty on the
```

the client.  A server MAY proceed with the connection without verifying the
token, though the server MUST NOT consider the client address validated.  If a
server chooses not to proceed with the handshake, it SHOULD immediately close
({{immediate-close}}) the connection with an INVALID_TOKEN error.  Note that a
server has not established any state for the connection at this point and so

> does not enter the closing period.

Reviewer 3: Would it make sense to change the last two sentences to something like: _If a server chooses not to proceed with the handshake, it SHOULD immediately send a CONNECTION_CLOSE frame with an INVALID_TOKEN error without creating any state. As the server does not establish any state for the connection, it does not enter the closing period ({{immediate-close}})._

Current text is confusing as the normative text suggests following the rule in {{immediate-close}}, then argues *as if it is a fact* that a server would not be retaining any state, even though what we want to do here is to *recommend* the server to send CONNECTION_CLOSE in a stateless manner.
